lib/ddtrace/profiling/scheduler.rb

```diff
@@ -0,0 +1,83 @@
+require 'ddtrace/utils/time'
+
+require 'ddtrace/worker'
+require 'ddtrace/workers/polling'
+
+module Datadog
+  module Profiling
+    # Pulls and exports profiling data on an async interval basis
+    class Scheduler < Worker
+      include Workers::Polling
+
+      DEFAULT_INTERVAL = 60
+      MIN_INTERVAL = 0
+
+      attr_reader \
+        :exporters,
+        :recorder
+
+      def initialize(recorder, exporters, options = {})
+        @recorder = recorder
+        @exporters = [exporters].flatten
+
+        # Workers::Async::Thread settings
+        # Restart in forks by default
+        self.fork_policy = options.fetch(:fork_policy, Workers::Async::Thread::FORK_POLICY_RESTART)
+
+        # Workers::IntervalLoop settings
+        self.loop_base_interval = options.fetch(:interval, DEFAULT_INTERVAL)
+
+        # Workers::Polling settings
+        self.enabled = options.fetch(:enabled, false)
+      end
+
+      def start
+        perform
+      end
+
+      def perform
+        flush_and_wait
+      end
+
+      def loop_back_off?
+        false
+      end
+
+      def after_fork
+        # Clear recorder's buffers by flushing events.
+        # Objects from parent process will copy-on-write,
+        # and we don't want to send events for the wrong process.
+        recorder.pop
+      end
+
+      def flush_and_wait
+        run_time = Datadog::Utils::Time.measure do
+          flush_events
+        end
+
+        # Update wait time to try to wake consistently on time.
+        # Don't drop below the minimum interval.
+        self.loop_wait_time = [loop_base_interval - run_time, MIN_INTERVAL].max
```

Comment:
Is it possible for `run_time` to be smaller then the last value of `loop_wait_time`?
If not, then `[loop_base_interval - run_time, MIN_INTERVAL].max` will always return `MIN_INTERVAL` here, is that intended?

Reply:
`run_time` is the monotonic clock measurement of how long it took to flush.

The intention of this line is adjust the wait time downwards from the base interval, to get an even interval between flushes. In the unlikely event that the flush took a very long time, the new wait time could be very short, even < 0. This `max` means to prevent the wait time from being too short.

Since our minimum wait time is 0, it prevents the wait time from going negative which would raise an error.